### PR TITLE
Feat/unify t2b t2kg

### PR DESCRIPTION
**Pull request by gurdeep330**
# For authors

## Description

https://github.com/user-attachments/assets/40395621-8d7e-4252-b145-a914c8dde80d

In this PR, I propose a unified agent that supervises T2B and T2KG. Here is the summary of the PR:
1. I have used the recently released library from Langgraph [langgraph-supervisor-py](https://github.com/langchain-ai/langgraph-supervisor-py) to design a main agent (Talk2AIAgents4Pharma) that supervises T2B and T2KG.
2. I have also created a new streamlit app that renders the main agent with minimal code change.

In the subsequent PRs, I will:
1. Add Jupyter notebooks
2. Bring T2S into the mix
3. Consolidate the streamlit apps code

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [X] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [X] My code follows the style guidelines mentioned in the Code/DevOps guides
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation (e.g. MkDocs)
- [X] My changes generate no new warnings
- [X] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [X] New and existing tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

File: app/frontend/streamlit_app_talk2aiagents4pharma.py [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+            key="text_embedding_model",
+            on_change=streamlit_utils.update_text_embedding_model,
+            kwargs={"app": app},
+            help="Used for Retrival Augmented Generation (RAG) and other tasks."
+        )
+
+        # T2B Upload files
+        uploaded_sbml_file = streamlit_utils.get_t2b_uploaded_files(app)
+
+        # T2KG Upload files
```
Comment (dmccloskey) [inline app/frontend/streamlit_app_talk2aiagents4pharma.py:185]: @gurdeep330 and @awmulyadi Perhaps not now but in a future PR, we need to come up with a better solution that the multitude of file uploads. Perhaps we could implement something similar to MS Teams (and other messaging apps) where you can add a file to the message itself. That way a user can either provide a tag to identify the file or the main agent can ask the user to specify what the file is/for if it is unclear from the context.

File: docs/index.md [hunk 2]
```diff
@@ -16,6 +16,7 @@ Our toolkit currently consists of the following agents:
 - **Talk2KnowledgeGraphs** _(v1 in progress)_: Access and explore complex biological knowledge graphs for insightful data connections.
 - **Talk2Scholars** _(v1 in progress)_: Get recommendations for articles related to your choice. Download, query, and write/retrieve them to your reference manager (currently supporting Zotero).
 - **Talk2Cells** _(v1 in progress)_: Query and analyze sequencing data with ease.
+- **Talk2AIAgents4Pharma** _(v1 in progress)_: Converse with all the agents above (currently supports T2B and t2KG)
```
Comment (dmccloskey) [inline docs/index.md:19]:
```suggestion
- **Talk2AIAgents4Pharma** _(v1 in progress)_: Converse with all the agents above (currently supports T2B and T2KG)
```

File: mkdocs.yml [hunk 3]
```diff
@@ -3,6 +3,13 @@ site_name: AIAgents4Pharma
 nav:
   - Home: index.md
 
+  - Talk2AIAgents4Pharma:
+      - Introduction: talk2aiagents4pharma/intro.md
+      - Agents:
+          - T2B agent: talk2aiagents4pharma/agents/main_agent.md
```
Comment (dmccloskey) [inline mkdocs.yml:9]: Is T2KG missing here?

File: .github/workflows/tests_talk2aiagents4pharma.yml [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+          source venv/bin/activate
+          pip install --upgrade pip
+          pip3 install --break-system-packages -r requirements.txt
+
+      # pylint
+      - name: Run pylint
+        working-directory: ${{ github.workspace }}
+        run: |
+          source venv/bin/activate
+          pylint --disable=R0801,R0902,W0221,W0122 aiagents4pharma/talk2aiagents4pharma
```
Comment (dmccloskey) [inline .github/workflows/tests_talk2aiagents4pharma.yml:66]: Nice work adapting this workflow to conditionally run on changes to Talk2AIAgents4Pharma, T2B, or T2KG and then just running the tests for Talk2AiAgents4Pharma.

File: aiagents4pharma/talk2aiagents4pharma/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  "plotly-express==0.4.1",
+  "seaborn==0.13.2",
+  "scanpy==1.11.0"
+]
+dynamic = ["version"]
+ 
+# Package discovery for the submodule
+[tool.setuptools]
+packages = [
+    "talk2biomodels",
```
[the remark this answers is not in the document]
Comment (gurdeep330) [reply aiagents4pharma/talk2aiagents4pharma/pyproject.toml:81]: Yes, even talk2biomodels should be omitted and replaced by talk2aiagents4pharma. Thanks for spotting 👍

File: mkdocs.yml (hunk 3 above)
Comment (gurdeep330) [reply mkdocs.yml:9]: Yes, even T2B should be omitted and replaced by T2AA4P. Thanks for spotting 👍

File: aiagents4pharma/talk2aiagents4pharma/configs/app/frontend/default.yaml
[the remark this answers is not in the document]
Comment (gurdeep330) [reply aiagents4pharma/talk2aiagents4pharma/configs/app/frontend/default.yaml:1]: I deleted the app config because it is identical to the one on T2KG. I am now loading it directly from T2KG's configs.
I think we still need the `agents` configs because this is unique to T2AA4P.

File: app/frontend/streamlit_app_talk2aiagents4pharma.py (hunk 1 above)
Comment (gurdeep330) [reply app/frontend/streamlit_app_talk2aiagents4pharma.py:184]: @dmccloskey That's a good suggestion. We will coordinate to set this up in the coming weeks. We also need to find a way to reduce redundancy in the streamlit code.

**Comment by gurdeep330** [comment]
Thanks for the suggestions and reviews, @dmccloskey

**Review by dmccloskey — APPROVED** [review]
Not bad for Sunday PR 😉. Your turn for review @awmulyadi

**Comment by gurdeep330** [comment]
@awmulyadi do you mind cloning this PR and testing out T2KG use cases? Thx!

File: app/frontend/streamlit_app_talk2aiagents4pharma.py [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+    st.session_state.topk_edges = cfg.reasoning_subgraph_topk_edges
+
+# Get the app
+app = st.session_state.app
+
+# Apply custom CSS
+streamlit_utils.apply_css()
+
+# Sidebar
+with st.sidebar:
```
[the remark this answers is not in the document]
Comment (awmulyadi) [reply app/frontend/streamlit_app_talk2aiagents4pharma.py:132]: @dmccloskey Yes, practically it is possible to perform these parameter extractions (top-k nodes and edges) via the user prompt to replace the default ones. However, there should be thorough test cases for every possible value of them. I opened issue #123 related to this feature request.

File: app/frontend/streamlit_app_talk2aiagents4pharma.py [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+from streamlit_feedback import streamlit_feedback
+from langchain_core.messages import SystemMessage, HumanMessage, AIMessage
+from langchain_core.messages import ChatMessage
+from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
+from utils import streamlit_utils
+
+st.set_page_config(page_title="Talk2AIAgents4Pharma",
+                   page_icon="🤖",
+                   layout="wide",
+                   initial_sidebar_state="collapsed",)
```
Comment (awmulyadi) [inline app/frontend/streamlit_app_talk2aiagents4pharma.py:21]: I set the `initial_sidebar_state` to be "collapsed" 👍

File: app/frontend/streamlit_app_talk2aiagents4pharma.py [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+        ("system", "Welcome to Talk2AIAgents4Pharma!"),
+        MessagesPlaceholder(variable_name='chat_history', optional=True),
+        ("human", "{input}"),
+        ("placeholder", "{agent_scratchpad}"),
+])
+
+# Initialize current user
+if "current_user" not in st.session_state:
+    st.session_state.current_user = cfg_t2kg.default_user
+
```
Comment (awmulyadi) [inline app/frontend/streamlit_app_talk2aiagents4pharma.py:71]: I included missing `current_user` in `st.session_state`.

**Review by awmulyadi — APPROVED** [review]
Hi @gurdeep330. This is a great update to seamlessly unify T2B and T2KG with minimal modification of each agent by using langgraph-supervisor. My only concern is that the current UI version has no Ollama options. And also, T2KG doesn't support the text-embedding options from the dropdown for its subgraph extraction process. We need to discuss this for subsequent PRs regarding this issue.

Note that I have committed my fix for adding missing `current_user` in the Streamlit app's `session_state` related to the uploading files functionality for T2K.

Other than that, all my prompts for T2KG work! Well done!
